File: services/api/src/api/services/workflow/validation.py
"""Semantic graph validation for workflow definitions — shared, structured rules.

The Pydantic contract (``schemas/workflow_definition.py``) guarantees a graph is
*structurally* well-formed (known types, unique ids, edges reference real nodes).
This module layers the *semantic* BPMN rules on top and returns a list of
:class:`Issue` keyed by node/edge so the API, the AI assistant, and the frontend
(which mirrors these rules in ``validation.ts``) can surface precise, actionable
feedback. ``error`` issues block publish; ``warning`` issues are advisory.

Checks run against the *normalized* graph (``compat.normalize``) so legacy
condition/switch/delay nodes are validated with the same rules as their BPMN
equivalents.
"""

from __future__ import annotations

from collections import defaultdict, deque
from dataclasses import dataclass
from typing import Any, Literal

from pydantic import ValidationError

from api.schemas.workflow_definition import WorkflowDefinitionModel, WorkflowNode
from api.services.workflow import compat
from api.services.workflow import constants as C

Severity = Literal["error", "warning"]


@dataclass(frozen=True)
class Issue:
    severity: Severity
    code: str
    message: str
    node_id: str | None = None
    edge_id: str | None = None

    def as_dict(self) -> dict[str, Any]:
        return {
            "severity": self.severity,
            "code": self.code,
            "message": self.message,
            "node_id": self.node_id,
            "edge_id": self.edge_id,
        }


def has_errors(issues: list[Issue]) -> bool:
    return any(i.severity == "error" for i in issues)


def validate_definition(
    definition: dict[str, Any] | WorkflowDefinitionModel | None,
) -> list[Issue]:
    """Return all semantic issues for a workflow graph (empty = clean).

    A structurally-malformed definition returns ``error`` issues describing the
    parse failures rather than raising, so a single call surfaces every problem.
    """
    try:
        raw = (
            definition
            if isinstance(definition, WorkflowDefinitionModel)
            else WorkflowDefinitionModel.parse(definition)
        )
    except ValidationError as exc:
        return [Issue("error", "malformed", _fmt_error(err)) for err in exc.errors()] or [
            Issue("error", "malformed", "definition is not a valid workflow graph")
        ]

    authored_v2 = raw.is_v2
    model = compat.normalize(raw)
    out_edges, incoming = _adjacency(model)

    issues: list[Issue] = []
    _check_triggers(model, issues)
    _check_reachability(model, out_edges, issues)
    _check_gateways(model, out_edges, incoming, issues)
    _check_boundary_events(model, issues)
    _check_event_based_gateways(model, out_edges, issues)
    _check_loops(model, out_edges, issues)
    if authored_v2:
        _check_end_presence(model, issues)
    return issues


# --------------------------------------------------------------------------- #
# helpers
# --------------------------------------------------------------------------- #
def _fmt_error(err: dict[str, Any]) -> str:
    loc = ".".join(str(p) for p in err.get("loc", ()))
    msg = err.get("msg", "invalid")
    return f"{loc}: {msg}" if loc else msg


def _adjacency(model: WorkflowDefinitionModel):
    out_edges: dict[str, list] = defaultdict(list)
    incoming: dict[str, int] = defaultdict(int)
    for edge in model.edges:
        out_edges[edge.source].append(edge)
        incoming[edge.target] += 1
    return out_edges, incoming


def _boundary_map(model: WorkflowDefinitionModel) -> dict[str, list[str]]:
    """activity id -> boundary event ids attached to it (a non-edge reachability
    link: a boundary event is reachable whenever its host activity is)."""
    by_host: dict[str, list[str]] = defaultdict(list)
    for node in model.nodes:
        if _is_boundary(node):
            host = node.data.get("attached_to")
            if host:
                by_host[host].append(node.id)
    return by_host


def _is_boundary(node: WorkflowNode) -> bool:
    return node.type == C.NODE_EVENT and node.data.get("position") == C.EVENT_BOUNDARY


def _is_catch_target(node: WorkflowNode | None) -> bool:
    if node is None:
        return False
    if node.type == C.NODE_EVENT and node.data.get("position") == C.EVENT_INTERMEDIATE:
        return node.data.get("throw_catch", "catch") == "catch"
    return node.type == C.NODE_TASK and node.task_type == C.TASK_RECEIVE


def _makes_progress(node: WorkflowNode) -> bool:
    """A node that does real work (a task) or waits (a timer/message/signal
    catch) — a loop containing one can't tight-spin."""
    if node.type == C.NODE_TASK:
        return True
    if node.type == C.NODE_EVENT:
        return node.data.get("event_type") in C.WAIT_EVENT_TYPES
    return False


# --------------------------------------------------------------------------- #
# checks
# --------------------------------------------------------------------------- #
def _check_triggers(model: WorkflowDefinitionModel, issues: list[Issue]) -> None:
    triggers = [n for n in model.nodes if n.type == C.NODE_TRIGGER]
    if not triggers:
        issues.append(
            Issue("error", "no-trigger", "The workflow has no trigger/start event, so it can never run.")
        )
    elif len(triggers) > 1:
        issues.append(
            Issue(
                "warning",
                "multiple-triggers",
                f"{len(triggers)} start events — a separate run starts from each.",
            )
        )


def _check_reachability(model: WorkflowDefinitionModel, out_edges, issues: list[Issue]) -> None:
    starts = [n.id for n in model.nodes if n.type == C.NODE_TRIGGER]
    boundary_by_host = _boundary_map(model)
    seen: set[str] = set()
    queue = deque(starts)
    while queue:
        current = queue.popleft()
        if current in seen:
            continue
        seen.add(current)
        for edge in out_edges.get(current, []):
            queue.append(edge.target)
        for boundary_id in boundary_by_host.get(current, []):
            queue.append(boundary_id)
    for node in model.nodes:
        if node.type == C.NODE_TRIGGER or node.id in seen:
            continue
        issues.append(
            Issue(
                "warning",
                "unreachable",
                f"Node {node.id!r} can't be reached from any start event and won't run.",
                node_id=node.id,
            )
        )


def _check_gateways(model: WorkflowDefinitionModel, out_edges, incoming, issues: list[Issue]) -> None:
    for node in model.nodes:
        if node.type != C.NODE_GATEWAY:
            continue
        gateway_type = node.gateway_type or C.GATEWAY_EXCLUSIVE
        out_count = len(out_edges.get(node.id, []))
        in_count = incoming.get(node.id, 0)
        if gateway_type in C.FORKING_GATEWAY_TYPES:
            if out_count < 2 and in_count < 2:
                issues.append(
                    Issue(
                        "warning",
                        "degenerate-gateway",
                        f"{gateway_type} gateway {node.id!r} neither forks (needs ≥2 outgoing) "
                        f"nor joins (needs ≥2 incoming).",
                        node_id=node.id,
                    )
                )
        elif gateway_type == C.GATEWAY_EXCLUSIVE and out_count >= 2:
            handles = [edge.source_handle for edge in out_edges.get(node.id, [])]
            branches_on_condition = node.data.get("expr") is not None or bool(node.data.get("cases"))
            if branches_on_condition and C.HANDLE_DEFAULT not in handles:
                issues.append(
                    Issue(
                        "warning",
                        "exclusive-no-default",
                        f"Exclusive gateway {node.id!r} has no default branch; a token matching no "
                        f"condition would stop here.",
                        node_id=node.id,
                    )
                )


def _check_boundary_events(model: WorkflowDefinitionModel, issues: list[Issue]) -> None:
    ids = {n.id: n for n in model.nodes}
    for node in model.nodes:
        if not _is_boundary(node):
            continue
        host_id = node.data.get("attached_to")
        if not host_id:
            issues.append(
                Issue(
                    "error",
                    "boundary-unattached",
                    f"Boundary event {node.id!r} is not attached to an activity.",
                    node_id=node.id,
                )
            )
            continue
        host = ids.get(host_id)
        if host is None:
            issues.append(
                Issue(
                    "error",
                    "boundary-bad-attach",
                    f"Boundary event {node.id!r} is attached to unknown node {host_id!r}.",
                    node_id=node.id,
                )
            )
        elif host.type != C.NODE_TASK:
            issues.append(
                Issue(
                    "warning",
                    "boundary-nonactivity",
                    f"Boundary event {node.id!r} should attach to a task, not a {host.type!r}.",
                    node_id=node.id,
                )
            )


def _check_event_based_gateways(model: WorkflowDefinitionModel, out_edges, issues: list[Issue]) -> None:
    ids = {n.id: n for n in model.nodes}
    for node in model.nodes:
        if node.type != C.NODE_GATEWAY or node.gateway_type != C.GATEWAY_EVENT_BASED:
            continue
        for edge in out_edges.get(node.id, []):
            if not _is_catch_target(ids.get(edge.target)):
                issues.append(
                    Issue(
                        "error",
                        "event-gateway-target",
                        f"Event-based gateway {node.id!r} must lead only to catch events or receive "
                        f"tasks; {edge.target!r} is not one.",
                        node_id=node.id,
                        edge_id=edge.id,
                    )
                )


def _check_loops(model: WorkflowDefinitionModel, out_edges, issues: list[Issue]) -> None:
    """Cycles are allowed (the token engine bounds them with a step budget), but a
    loop with no task/wait node can spin without progress — warn on those."""
    for scc in _strongly_connected(model, out_edges):
        if len(scc) == 1:
            node_id = next(iter(scc))
            self_loop = any(e.target == node_id for e in out_edges.get(node_id, []))
            if not self_loop:
                continue
        ids = {n.id: n for n in model.nodes}
        if not any(_makes_progress(ids[nid]) for nid in scc if nid in ids):
            issues.append(
                Issue(
                    "warning",
                    "loop-no-progress",
                    f"Loop through {sorted(scc)} contains no task or wait step and may spin without "
                    f"making progress.",
                    node_id=sorted(scc)[0],
                )
            )


def _check_end_presence(model: WorkflowDefinitionModel, issues: list[Issue]) -> None:
    has_end = any(
        n.type == C.NODE_EVENT and n.data.get("position") == C.EVENT_END for n in model.nodes
    )
    if not has_end:
        issues.append(
            Issue("warning", "no-end-event", "No explicit end event; add one so the flow's completion reads clearly.")
        )


def _strongly_connected(model: WorkflowDefinitionModel, out_edges) -> list[set[str]]:
    """Tarjan's SCC over the node graph (iterative — safe for any graph size)."""
    index_of: dict[str, int] = {}
    low: dict[str, int] = {}
    on_stack: set[str] = set()
    stack: list[str] = []
    sccs: list[set[str]] = []
    counter = 0
    node_ids = [n.id for n in model.nodes]

    for root in node_ids:
        if root in index_of:
            continue
        work: list[tuple[str, int]] = [(root, 0)]
        while work:
            node, child_idx = work[-1]
            if child_idx == 0:
                index_of[node] = low[node] = counter
                counter += 1
                stack.append(node)
                on_stack.add(node)
            targets = out_edges.get(node, [])
            if child_idx < len(targets):
                work[-1] = (node, child_idx + 1)
                nxt = targets[child_idx].target
                if nxt not in index_of:
                    work.append((nxt, 0))
                elif nxt in on_stack:
                    low[node] = min(low[node], index_of[nxt])
            else:
                if low[node] == index_of[node]:
                    component: set[str] = set()
                    while True:
                        w = stack.pop()
                        on_stack.discard(w)
                        component.add(w)
                        if w == node:
                            break
                    sccs.append(component)
                work.pop()
                if work:
                    parent = work[-1][0]
                    low[parent] = min(low[parent], low[node])
    return sccs

File: ui/src/components/workflows/validation.ts
/**
 * Semantic workflow-graph validation — the frontend mirror of the backend
 * authority `services/api/src/api/services/workflow/validation.py`. Codes,
 * severities and reachability semantics match so the designer surfaces the same
 * precise, actionable feedback the API would return on save/publish.
 *
 * Like the backend, checks run against a *normalized* graph (legacy
 * condition/switch/delay/action nodes mapped to their BPMN category) so old and
 * new graphs are validated by one set of rules. Cycles are ALLOWED (the token
 * engine bounds them with a step budget); only a loop with no task/wait step
 * that could spin is warned about.
 */
import type { GraphEdge, GraphNode, WorkflowDefinition } from "@/lib/api/workflows";

import {
  FORKING_GATEWAY_TYPES,
  GATEWAY_TYPES,
  HANDLE_DEFAULT,
  NODE_EVENT,
  NODE_GATEWAY,
  NODE_TASK,
  NODE_TRIGGER,
  SCHEMA_VERSION,
  TASK_TYPES,
  EVENT_POSITIONS,
  EVENT_TYPES,
  WAIT_EVENT_TYPES,
  type GatewayType,
} from "./nodes/nodeMeta";

export type Severity = "error" | "warning";

export interface Issue {
  severity: Severity;
  code: string;
  message: string;
  nodeId?: string;
  edgeId?: string;
}

export function hasErrors(issues: Issue[]): boolean {
  return issues.some((i) => i.severity === "error");
}

// --------------------------------------------------------------------------- //
// Normalization (mirror compat.normalize)
// --------------------------------------------------------------------------- //
interface NormNode {
  id: string;
  type: string;
  data: Record<string, unknown>;
}

const SEND_ACTIONS = new Set(["send_email", "send_webhook", "send_form"]);
const GATEWAY_LEGACY = new Set(["condition", "switch", "merge", "passthrough"]);
const ALL_NODE_TYPES = new Set<string>([
  NODE_TRIGGER,
  NODE_TASK,
  NODE_GATEWAY,
  NODE_EVENT,
  "action",
  "condition",
  "switch",
  "delay",
  "merge",
  "passthrough",
]);
const NODE_ID_RE = /^[A-Za-z0-9_-]{1,64}$/;

function normalizeNode(n: GraphNode): NormNode {
  const data: Record<string, unknown> = { ...(n.data ?? {}) };
  if (n.type === "action") {
    if (data.task_type === undefined)
      data.task_type = SEND_ACTIONS.has(String(data.action_type)) ? "send" : "service";
    return { id: n.id, type: NODE_TASK, data };
  }
  if (GATEWAY_LEGACY.has(n.type)) {
    if (data.gateway_type === undefined) data.gateway_type = "exclusive";
    return { id: n.id, type: NODE_GATEWAY, data };
  }
  if (n.type === "delay") {
    if (data.position === undefined) data.position = "intermediate";
    if (data.event_type === undefined) data.event_type = "timer";
    if (data.throw_catch === undefined) data.throw_catch = "catch";
    return { id: n.id, type: NODE_EVENT, data };
  }
  return { id: n.id, type: n.type, data: data };
}

/** True when a graph should be treated as v2 (mirrors WorkflowDefinitionModel.is_v2). */
function isAuthoredV2(def: WorkflowDefinition): boolean {
  if ((def.schema_version ?? 1) >= SCHEMA_VERSION) return true;
  return (def.nodes ?? []).some(
    (n) => n.type === NODE_TASK || n.type === NODE_GATEWAY || n.type === NODE_EVENT,
  );
}

// --------------------------------------------------------------------------- //
// Structural gate (mirror the Pydantic contract -> "malformed" issues)
// --------------------------------------------------------------------------- //
function structuralIssues(def: WorkflowDefinition): Issue[] {
  const issues: Issue[] = [];
  const seen = new Set<string>();
  for (const n of def.nodes ?? []) {
    if (!NODE_ID_RE.test(n.id)) {
      issues.push({ severity: "error", code: "malformed", message: `node id ${JSON.stringify(n.id)} must match [A-Za-z0-9_-]{1,64}`, nodeId: n.id });
    }
    if (!ALL_NODE_TYPES.has(n.type)) {
      issues.push({ severity: "error", code: "malformed", message: `unknown node type ${JSON.stringify(n.type)}`, nodeId: n.id });
    }
    if (seen.has(n.id)) {
      issues.push({ severity: "error", code: "malformed", message: `duplicate node id ${JSON.stringify(n.id)}`, nodeId: n.id });
    }
    seen.add(n.id);
    subtypeIssue(n, issues);
  }
  for (const e of def.edges ?? []) {
    if (!seen.has(e.source) || !seen.has(e.target)) {
      issues.push({ severity: "error", code: "malformed", message: `Edge ${JSON.stringify(e.id ?? `${e.source}->${e.target}`)} references a node that no longer exists.`, edgeId: e.id ?? undefined });
    }
  }
  return issues;
}

function subtypeIssue(n: GraphNode, issues: Issue[]): void {
  const data = n.data ?? {};
  if (n.type === NODE_TASK && data.task_type != null && !(TASK_TYPES as readonly string[]).includes(String(data.task_type))) {
    issues.push({ severity: "error", code: "malformed", message: `task node ${JSON.stringify(n.id)}: unknown task_type ${JSON.stringify(data.task_type)}`, nodeId: n.id });
  } else if (n.type === NODE_GATEWAY && data.gateway_type != null && !(GATEWAY_TYPES as readonly string[]).includes(String(data.gateway_type))) {
    issues.push({ severity: "error", code: "malformed", message: `gateway node ${JSON.stringify(n.id)}: unknown gateway_type ${JSON.stringify(data.gateway_type)}`, nodeId: n.id });
  } else if (n.type === NODE_EVENT) {
    if (data.position != null && !(EVENT_POSITIONS as readonly string[]).includes(String(data.position))) {
      issues.push({ severity: "error", code: "malformed", message: `event node ${JSON.stringify(n.id)}: unknown position ${JSON.stringify(data.position)}`, nodeId: n.id });
    }
    if (data.event_type != null && !(EVENT_TYPES as readonly string[]).includes(String(data.event_type))) {
      issues.push({ severity: "error", code: "malformed", message: `event node ${JSON.stringify(n.id)}: unknown event_type ${JSON.stringify(data.event_type)}`, nodeId: n.id });
    }
  }
}

// --------------------------------------------------------------------------- //
// Graph helpers
// --------------------------------------------------------------------------- //
type Adjacency = { outEdges: Map<string, GraphEdge[]>; incoming: Map<string, number> };

function adjacency(nodes: NormNode[], edges: GraphEdge[]): Adjacency {
  const outEdges = new Map<string, GraphEdge[]>();
  const incoming = new Map<string, number>();
  for (const n of nodes) outEdges.set(n.id, []);
  for (const e of edges) {
    outEdges.get(e.source)?.push(e);
    incoming.set(e.target, (incoming.get(e.target) ?? 0) + 1);
  }
  return { outEdges, incoming };
}

function isBoundary(n: NormNode): boolean {
  return n.type === NODE_EVENT && n.data.position === "boundary";
}

function boundaryMap(nodes: NormNode[]): Map<string, string[]> {
  const byHost = new Map<string, string[]>();
  for (const n of nodes) {
    if (!isBoundary(n)) continue;
    const host = n.data.attached_to;
    if (typeof host === "string") {
      const list = byHost.get(host) ?? [];
      list.push(n.id);
      byHost.set(host, list);
    }
  }
  return byHost;
}

function isCatchTarget(n: NormNode | undefined): boolean {
  if (!n) return false;
  if (n.type === NODE_EVENT && n.data.position === "intermediate") {
    return (n.data.throw_catch ?? "catch") === "catch";
  }
  return n.type === NODE_TASK && n.data.task_type === "receive";
}

function makesProgress(n: NormNode): boolean {
  if (n.type === NODE_TASK) return true;
  if (n.type === NODE_EVENT) return (WAIT_EVENT_TYPES as readonly string[]).includes(String(n.data.event_type));
  return false;
}

// --------------------------------------------------------------------------- //
// Checks
// --------------------------------------------------------------------------- //
function checkTriggers(nodes: NormNode[], issues: Issue[]): void {
  const triggers = nodes.filter((n) => n.type === NODE_TRIGGER);
  if (triggers.length === 0) {
    issues.push({ severity: "error", code: "no-trigger", message: "The workflow has no trigger/start event, so it can never run." });
  } else if (triggers.length > 1) {
    issues.push({ severity: "warning", code: "multiple-triggers", message: `${triggers.length} start events — a separate run starts from each.` });
  }
}

function checkReachability(nodes: NormNode[], out: Adjacency, issues: Issue[]): void {
  const starts = nodes.filter((n) => n.type === NODE_TRIGGER).map((n) => n.id);
  const byHost = boundaryMap(nodes);
  const seen = new Set<string>();
  const queue = [...starts];
  while (queue.length > 0) {
    const current = queue.shift() as string;
    if (seen.has(current)) continue;
    seen.add(current);
    for (const e of out.outEdges.get(current) ?? []) queue.push(e.target);
    for (const b of byHost.get(current) ?? []) queue.push(b);
  }
  for (const n of nodes) {
    if (n.type === NODE_TRIGGER || seen.has(n.id)) continue;
    issues.push({ severity: "warning", code: "unreachable", message: `Node ${JSON.stringify(n.id)} can't be reached from any start event and won't run.`, nodeId: n.id });
  }
}

function checkGateways(nodes: NormNode[], adj: Adjacency, issues: Issue[]): void {
  for (const n of nodes) {
    if (n.type !== NODE_GATEWAY) continue;
    const gatewayType = ((n.data.gateway_type as GatewayType) || "exclusive") as GatewayType;
    const outCount = (adj.outEdges.get(n.id) ?? []).length;
    const inCount = adj.incoming.get(n.id) ?? 0;
    if ((FORKING_GATEWAY_TYPES as readonly string[]).includes(gatewayType)) {
      if (outCount < 2 && inCount < 2) {
        issues.push({ severity: "warning", code: "degenerate-gateway", message: `${gatewayType} gateway ${JSON.stringify(n.id)} neither forks (needs ≥2 outgoing) nor joins (needs ≥2 incoming).`, nodeId: n.id });
      }
    } else if (gatewayType === "exclusive" && outCount >= 2) {
      const handles = (adj.outEdges.get(n.id) ?? []).map((e) => e.source_handle);
      const branchesOnCondition = n.data.expr != null || Boolean(n.data.cases);
      if (branchesOnCondition && !handles.includes(HANDLE_DEFAULT)) {
        issues.push({ severity: "warning", code: "exclusive-no-default", message: `Exclusive gateway ${JSON.stringify(n.id)} has no default branch; a token matching no condition would stop here.`, nodeId: n.id });
      }
    }
  }
}

function checkBoundaryEvents(nodes: NormNode[], issues: Issue[]): void {
  const byId = new Map(nodes.map((n) => [n.id, n]));
  for (const n of nodes) {
    if (!isBoundary(n)) continue;
    const host = n.data.attached_to;
    if (typeof host !== "string" || host === "") {
      issues.push({ severity: "error", code: "boundary-unattached", message: `Boundary event ${JSON.stringify(n.id)} is not attached to an activity.`, nodeId: n.id });
      continue;
    }
    const hostNode = byId.get(host);
    if (!hostNode) {
      issues.push({ severity: "error", code: "boundary-bad-attach", message: `Boundary event ${JSON.stringify(n.id)} is attached to unknown node ${JSON.stringify(host)}.`, nodeId: n.id });
    } else if (hostNode.type !== NODE_TASK) {
      issues.push({ severity: "warning", code: "boundary-nonactivity", message: `Boundary event ${JSON.stringify(n.id)} should attach to a task, not a ${JSON.stringify(hostNode.type)}.`, nodeId: n.id });
    }
  }
}

function checkEventBasedGateways(nodes: NormNode[], adj: Adjacency, issues: Issue[]): void {
  const byId = new Map(nodes.map((n) => [n.id, n]));
  for (const n of nodes) {
    if (n.type !== NODE_GATEWAY || n.data.gateway_type !== "event_based") continue;
    for (const e of adj.outEdges.get(n.id) ?? []) {
      if (!isCatchTarget(byId.get(e.target))) {
        issues.push({ severity: "error", code: "event-gateway-target", message: `Event-based gateway ${JSON.stringify(n.id)} must lead only to catch events or receive tasks; ${JSON.stringify(e.target)} is not one.`, nodeId: n.id, edgeId: e.id ?? undefined });
      }
    }
  }
}

function checkLoops(nodes: NormNode[], adj: Adjacency, issues: Issue[]): void {
  const byId = new Map(nodes.map((n) => [n.id, n]));
  for (const scc of stronglyConnected(nodes, adj)) {
    if (scc.size === 1) {
      const only = [...scc][0];
      const selfLoop = (adj.outEdges.get(only) ?? []).some((e) => e.target === only);
      if (!selfLoop) continue;
    }
    const progress = [...scc].some((id) => {
      const node = byId.get(id);
      return node ? makesProgress(node) : false;
    });
    if (!progress) {
      const sorted = [...scc].sort();
      issues.push({ severity: "warning", code: "loop-no-progress", message: `Loop through ${JSON.stringify(sorted)} contains no task or wait step and may spin without making progress.`, nodeId: sorted[0] });
    }
  }
}

function checkEndPresence(nodes: NormNode[], issues: Issue[]): void {
  const hasEnd = nodes.some((n) => n.type === NODE_EVENT && n.data.position === "end");
  if (!hasEnd) {
    issues.push({ severity: "warning", code: "no-end-event", message: "No explicit end event; add one so the flow's completion reads clearly." });
  }
}

/** Tarjan's SCC (iterative — safe for any graph size), mirroring the backend. */
function stronglyConnected(nodes: NormNode[], adj: Adjacency): Set<string>[] {
  const indexOf = new Map<string, number>();
  const low = new Map<string, number>();
  const onStack = new Set<string>();
  const stack: string[] = [];
  const sccs: Set<string>[] = [];
  let counter = 0;

  for (const root of nodes.map((n) => n.id)) {
    if (indexOf.has(root)) continue;
    const work: [string, number][] = [[root, 0]];
    while (work.length > 0) {
      const [node, childIdx] = work[work.length - 1];
      if (childIdx === 0) {
        indexOf.set(node, counter);
        low.set(node, counter);
        counter += 1;
        stack.push(node);
        onStack.add(node);
      }
      const targets = adj.outEdges.get(node) ?? [];
      if (childIdx < targets.length) {
        work[work.length - 1] = [node, childIdx + 1];
        const next = targets[childIdx].target;
        if (!indexOf.has(next)) {
          work.push([next, 0]);
        } else if (onStack.has(next)) {
          low.set(node, Math.min(low.get(node) ?? 0, indexOf.get(next) ?? 0));
        }
      } else {
        if ((low.get(node) ?? 0) === (indexOf.get(node) ?? 0)) {
          const component = new Set<string>();
          for (;;) {
            const w = stack.pop() as string;
            onStack.delete(w);
            component.add(w);
            if (w === node) break;
          }
          sccs.push(component);
        }
        work.pop();
        if (work.length > 0) {
          const parent = work[work.length - 1][0];
          low.set(parent, Math.min(low.get(parent) ?? 0, low.get(node) ?? 0));
        }
      }
    }
  }
  return sccs;
}

// --------------------------------------------------------------------------- //
// Public entry point
// --------------------------------------------------------------------------- //
/** Return every semantic issue for a workflow graph (empty = clean). */
export function validateGraph(def: WorkflowDefinition | null | undefined): Issue[] {
  if (!def) return [{ severity: "error", code: "malformed", message: "definition is not a valid workflow graph" }];

  const structural = structuralIssues(def);
  if (structural.length > 0) return structural;

  const authoredV2 = isAuthoredV2(def);
  const nodes = (def.nodes ?? []).map(normalizeNode);
  const adj = adjacency(nodes, def.edges ?? []);

  const issues: Issue[] = [];
  checkTriggers(nodes, issues);
  checkReachability(nodes, adj, issues);
  checkGateways(nodes, adj, issues);
  checkBoundaryEvents(nodes, issues);
  checkEventBasedGateways(nodes, adj, issues);
  checkLoops(nodes, adj, issues);
  if (authoredV2) checkEndPresence(nodes, issues);
  return issues;
}
